File: python/balanced_binary_tree.py
#!/usr/bin/env python

__author__ = 'Rio'

def is_balanced(root):
    result = check_height(root)
    if result == -1:
        return False
    return True
    
def check_height(root):
    if not root:
        return 0

    left = check_height(root.left)
    if left == -1:
        return -1

    right = check_height(root.right)
    if right == -1:
        return -1

    if abs(left - right) > 1:
        return -1

    return max(left, right) + 1

File: java/BalancedBinaryTree.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) { val = x; }
}
public class BalancedBinaryTree {
    public boolean isBalanced(TreeNode root) {
        return checkHeight(root) >= 0;
    }

    private int checkHeight(TreeNode node) {
        if (node == null) return 0;

        int left = checkHeight(node.left);
        int right = checkHeight(node.right);
        if (left < 0 || right < 0 || Math.abs(left - right) < 0) {
            return -1;
        }

        return Math.max(left, right) + 1;
    }
}
